libs/pilight/protocols/433.92/logilink_switch2.c

```diff
@@ -0,0 +1,234 @@
+	version.
+	pilight is distributed in the hope that it will be useful, but WITHOUT ANY
+	WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+	A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+	You should have received a copy of the GNU General Public License
+	along with pilight. If not, see	<http://www.gnu.org/licenses/>
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+
+#include "../../core/pilight.h"
+#include "../../core/common.h"
+#include "../../core/dso.h"
+#include "../../core/log.h"
+#include "../protocol.h"
+#include "../../core/binary.h"
+#include "../../core/gc.h"
+#include "logilink_switch2.h"
+
+#define PULSE_MULTIPLIER	3
+#define MIN_PULSE_LENGTH	279
+#define MAX_PULSE_LENGTH	289
+#define AVG_PULSE_LENGTH	284
+#define RAW_LENGTH				50
+
+static int validate(void) {
+	if(logilink_switch2->rawlen == RAW_LENGTH) {
+		if(logilink_switch2->raw[logilink_switch2->rawlen-1] >= (MIN_PULSE_LENGTH*PULSE_DIV) &&
+		   logilink_switch2->raw[logilink_switch2->rawlen-1] <= (MAX_PULSE_LENGTH*PULSE_DIV)) {
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+static void createMessage(int systemcode, int unitcode, int state) {
+	logilink_switch2->message = json_mkobject();
+	json_append_member(logilink_switch2->message, "systemcode", json_mknumber(systemcode, 0));
+	json_append_member(logilink_switch2->message, "unitcode", json_mknumber(unitcode, 0));
+	if(state == 0) {
+		json_append_member(logilink_switch2->message, "state", json_mkstring("on"));
+	} else {
+		json_append_member(logilink_switch2->message, "state", json_mkstring("off"));
+	}
+} 
+static void parseCode(void) {
+	int i = 0, x = 0, binary[RAW_LENGTH/2];
+	int systemcode = 0, state = 0, unitcode = 0;
+
+	for(x=0;x<logilink_switch2->rawlen-1;x+=2) {
+		if(logilink_switch2->raw[x] > (int)((double)AVG_PULSE_LENGTH*((double)PULSE_MULTIPLIER/2))) {
+			binary[i++] = 1;
+		} else {
+			binary[i++] = 0;
+		}
+	}
+	systemcode = binToDecRev(binary, 0, 19);
+	state = binary[23];
+	unitcode = binToDecRev(binary, 20, 22);
+
+	createMessage(systemcode, unitcode, state);
+}
+
+static void createLow(int s, int e) {
+	int i = 0;
+
+	for(i=s;i<=e;i+=2) {
+		logilink_switch2->raw[i]=(AVG_PULSE_LENGTH);
+		logilink_switch2->raw[i+1]=(PULSE_MULTIPLIER*AVG_PULSE_LENGTH);
+	}
+}
+
+static void createHigh(int s, int e) {
+	int i = 0;
+
+	for(i=s;i<=e;i+=2) {
+		logilink_switch2->raw[i]=(PULSE_MULTIPLIER*AVG_PULSE_LENGTH);
+		logilink_switch2->raw[i+1]=(AVG_PULSE_LENGTH);
+	}
+}
+
+static void clearCode(void) {
+	createLow(0, logilink_switch2->rawlen-2);
+}
+
+static void createSystemCode(int systemcode) {
+	int binary[255];
+	int length=0;
+	int i = 0, x = 38;
+
+	length = decToBin(systemcode, binary);
+	for(i=length;i>=0;i--) {
+		if(binary[i] == 1) {
+			createHigh(x, x+1);
+		}
+
+		x -= 2;
+	}
+}
+
+static void createUnitCode(int unitcode) {
+	switch(unitcode) {
+		case 7:
+			createHigh(40, 45);	// Button 1

```

Comment:
Can you check the alignment here?

Reply:
Did you mean the space behind createHigh(40, 45);?

Reply:
1. Can you fix your commit so it's properly linked to your account?
-> What do you mean with fixing the commit and linking it to my account? 
2. I also miss CMakeExclude.txt entry.
-> done
3. Can you make a PR for the rewrite branch as well including the unit test?
-> Where do I find the unit tests?

Reply:
> Did you mean the space behind createHigh(40, 45);

Yes